File: RxTools-library/src/main/java/com/vondear/rxtools/view/progress/sprite/CircleLayoutContainer.java
package com.vondear.rxtools.view.progress.sprite;

import android.graphics.Canvas;
import android.graphics.Rect;

/**
 * Created by ybq.
 */
public abstract class CircleLayoutContainer extends SpriteContainer {

    @Override
    public void drawChild(Canvas canvas) {
        for (int i = 0; i < getChildCount(); i++) {
            Sprite sprite = getChildAt(i);
            int count = canvas.save();
            canvas.rotate(i * 360 / getChildCount(),
                    getBounds().centerX(),
                    getBounds().centerY());
            sprite.draw(canvas);
            canvas.restoreToCount(count);
        }
    }

    @Override
    protected void onBoundsChange(Rect bounds) {
        super.onBoundsChange(bounds);
        bounds = clipSquare(bounds);
        int radius = (int) (bounds.width() * Math.PI / 3.6f / getChildCount());
        int left = bounds.centerX() - radius;
        int right = bounds.centerX() + radius;
        for (int i = 0; i < getChildCount(); i++) {
            Sprite sprite = getChildAt(i);
            sprite.setDrawBounds(left, bounds.top, right, bounds.top + radius * 2);
        }
    }
}

File: RxUI/src/main/java/com/tamsiree/rxui/view/progressing/sprite/CircleLayoutContainer.kt
package com.tamsiree.rxui.view.progressing.sprite

import android.graphics.Canvas
import android.graphics.Rect

/**
 * @author tamsiree
 */
abstract class CircleLayoutContainer : SpriteContainer() {
    override fun drawChild(canvas: Canvas) {
        for (i in 0 until childCount) {
            val sprite = getChildAt(i)
            val count = canvas.save()
            canvas.rotate(i * 360 / childCount.toFloat(),
                    bounds.centerX().toFloat(),
                    bounds.centerY().toFloat())
            sprite!!.draw(canvas)
            canvas.restoreToCount(count)
        }
    }

    override fun onBoundsChange(bounds: Rect) {
        var bounds = bounds
        super.onBoundsChange(bounds)
        bounds = clipSquare(bounds)
        val radius = (bounds.width() * Math.PI / 3.6f / childCount).toInt()
        val left = bounds.centerX() - radius
        val right = bounds.centerX() + radius
        for (i in 0 until childCount) {
            val sprite = getChildAt(i)
            sprite!!.setDrawBounds0(left, bounds.top, right, bounds.top + radius * 2)
        }
    }
}